File: Assignment 1/Heapsort/heapsort.py
def heapify(array):
    for i in range(1, len(array)):
        j = i
        while j > 0 and array[j] > array[(j - 1) // 2]:
            array[j], array[(j - 1) // 2] = array[(j - 1) // 2], array[j]
            j = (j - 1) // 2


def heapsort(array):
    heapify(array)

    for k in range(len(array) - 1, 0, -1):
        array[0], array[k] = array[k], array[0]

        j = 0
        while True:
            largest = j
            left = 2 * j + 1
            right = 2 * j + 2

            if left < k and array[left] > array[largest]:
                largest = left
            if right < k and array[right] > array[largest]:
                largest = right
            if largest == j:
                break

            array[j], array[largest] = array[largest], array[j]
            j = largest


if __name__ == '__main__':
    arr = [49, 27, 40, 59, 13, 55, 38, 2, 62, 48, 37, 4, 44, 22, 3, 51]
    heapsort(arr)
    print(arr)

File: Assignment 1/Heapsort/heapsort.cpp
#include <iostream>

void swap(int& a, int& b)
{
    int temp {a};
    a = b;
    b = temp;
}

void heapify(int* arr, int n)
{
    for(int i {1}; i < n; ++i)
    {
        int j {i};
        while(j > 0 && arr[j] > arr[(j - 1) / 2])
        {
            swap(arr[j], arr[(j - 1) / 2]);
            j = (j - 1) / 2;
        }
    }
}

void heapsort(int* arr, int n)
{
    heapify(arr, n);

    for(int k {n - 1}; k > 0; --k)
    {
        swap(arr[0], arr[k]);

        int largest, left, right;

        int j {0};
        while(true)
        {
            largest = j;
            left  = 2 * j + 1;
            right = 2 * j + 2;

            if(left < k && arr[left] > arr[largest])
            {
                largest = left;
            }
            if(right < k && arr[right] > arr[largest])
            {
                largest = right;
            }
            if(largest == j)
            {
                break;
            }

            swap(arr[j], arr[largest]);
            j = largest;
        }
    }
}

int main()
{
    int* arr {new int[16] {49, 27, 40, 59, 13, 55, 38, 2, 62, 48, 37, 4, 44, 22, 3, 51}};
    heapsort(arr, 16);

    std::cout << '[';
    for(int i {0}; i < 15; ++i)
    {
        std::cout << arr[i] << ", ";
    }
    std::cout << arr[15] << ']' << std::endl;

    delete[] arr;
    return 0;
}
